data-analytics/iot-nirvana/client/src/main/java/com/google/cloud/demo/iot/nirvana/client/MessagePublisher.java

```diff
@@ -0,0 +1,194 @@
+      String mqttClientId = generateMqttClientId();
+      MqttClient client = new MqttClient(mqttServerAddress, mqttClientId, new MemoryPersistence());
+      client.connect(connectOptions);
+      return client;
+    } catch (MqttException e) {
+      String message =
+          String.format("Failed to connect to the IoT server. Cause: %s", e.getMessage());
+      logError(message);
+      throw new ClientException(message, e);
+    }
+  }
+
+  /** Build the endpoint of the MQTT Cloud IoT Core server. */
+  protected String generateMqttServerAddress() {
+    return String.format(
+        "ssl://%s:%s", CLOUD_IOT_CORE_MQTT_BRIDGE_HOST_NAME, CLOUD_IOT_CORE_MQTT_BRIDGE_PORT);
+  }
+
+  /** Create the MQTT identifier of the device sending messages. */
+  protected String generateMqttClientId() throws ClientException {
+    return String.format(
+        "projects/%s/locations/%s/registries/%s/devices/%s",
+        getOptions().getGcpProjectId(),
+        getOptions().getGcpRegion(),
+        getOptions().getRegistryName(),
+        getDeviceId());
+  }
+
+  /**
+   * Create a JWT to authenticate this device. The device will be disconnected after the token
+   * expires, and will have to reconnect with a new token. The audience field should always be set
+   * to the GCP project id.
+   */
+  private String createJwtRsa() throws ClientException {
+    try {
+      DateTime now = new DateTime();
+      JwtBuilder jwtBuilder =
+          Jwts.builder()
+              .setIssuedAt(now.toDate())
+              .setExpiration(now.plusMinutes(20).toDate())
+              .setAudience(getOptions().getGcpProjectId());
+      byte[] keyBytes = Files.readAllBytes(Paths.get(getOptions().getPrivateKey()));
+      PKCS8EncodedKeySpec spec = new PKCS8EncodedKeySpec(keyBytes);
+      KeyFactory kf = KeyFactory.getInstance("RSA");
+      return jwtBuilder.signWith(SignatureAlgorithm.RS256, kf.generatePrivate(spec)).compact();
+    } catch (IOException e) {
+      String message = String.format("Cannot create JWT. Cause: %s", e.getMessage());
+      logError(message);
+      throw new ClientException(message, e);
+    } catch (NoSuchAlgorithmException e) {
+      String message = String.format("Cannot create JWT. Cause: %s", e.getMessage());
+      logError(message);
+      throw new ClientException(message, e);
+    } catch (InvalidKeySpecException e) {
+      String message = String.format("Cannot create JWT. Cause: %s", e.getMessage());
+      logError(message);
+      throw new ClientException(message, e);
+    }
+  }
+
+  /** Create the MQTT topic on which the device will publish messages to Cloud IoT Core. */
+  protected String generateMqttTopic() {
+    return String.format("/devices/%s/events", getDeviceId());
+  }
+
+  /** Log a debug message. */
+  private void logDebug(String message) throws ClientException {
+    LogUtils.writeLog(
+        logging,
+        message,
+        Severity.INFO,
```

Comment:
DEBUG?

Reply:
Changed in commit aa68e7f20fc4feb6103d063226e7caeb3f4cc266